### PR TITLE
[oneDPL] Create a subsection for is_execution_policy type trait

### PR DESCRIPTION
**Pull request by akukanov**
This PR describes oneDPL `is_execution_policy` trait in more detail, and also addresses https://github.com/uxlfoundation/oneAPI-spec/issues/558.

### CODE REVIEW

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (danhoeflinger) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:227]:
```suggestion
   Those in ``namespace oneapi::dpl::execution`` are provided for backward compatibility and may be deprecated in the future.
```

Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst]: Applied.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (danhoeflinger) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:194]: Considering we are encouraging people to use the version at `oneapi::dpl::`, should we list that one before going one level deeper into `oneapi::dpl::execution::`?

At a quick glance, a user may just use the first version which is listed, and not read the note below.

Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:194]: Why not; reordered.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:222]:
```suggestion
from its template argument.
```
It not a direct suggestion, it's a question. Is it a part of the spec? I would just drop it, honestly

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:228]:
```suggestion
   For a new code, use of the type trait utilities in ``namespace oneapi::dpl`` is strongly recommended. Those
```
Or "the"

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:186]:
```suggestion
oneDPL provides type trait utilities to detect its execution policy types at compile-time for the purpose of
```
Just a stylistic change. I feel it sounds more natural English-wise. Native speakers could help here :) And also not sure whether "compile-time" should be spelled via "-". That's how I usually spell it and how cppreference spells it

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:228]: Do you think "code" is countable? :) To me, it's like "water" or "air" :)

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:186]: Thanks, I will apply. The hyphen is not needed here, but would be needed if it was a compound adjective, like in "compile-time dispatch".

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:222]: > Is it a part of the spec?

Which spec do you mean? If C++ - no, there is nothing said about stripping references etc.
My intent is to give some usage hint for `is_execution_policy` as it might not be obvious that in practice it requires `decay`, `remove_cvref` etc. to work correctly.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:228]: Between the two options, I prefer the original.  It could also be stated like 
"When writing new code,... "

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (danhoeflinger) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:222]: Optional: It is a little strange to have this inline note directly followed by a multi-line note.
I think it may be better to either remove the note descriptor and wrapping and just tag on to the previous sentence: " and is best used in combination with ``std::decay_t<T>'' or similar type transformation utilities"
 -or-
 We can remove this note from the spec in favor of some similar advice in the oneDPL documentation.  We may be able to assume some level of sophistication of someone reading this specification, where the previous sentence implies this note.

**Review by danhoeflinger — APPROVED** [review]
Generally LGTM,  optional comment about the decay note.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst]: I reordered things in the subsection, which addressed this concern.

**Review by danhoeflinger — APPROVED** [review]
LGTM

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:200]:
```suggestion
      inline constexpr bool is_execution_policy_v = oneapi::dpl::is_execution_policy<T>::value;
```

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:208]:
```suggestion
        inline constexpr bool is_execution_policy_v = oneapi::dpl::execution::is_execution_policy<T>::value;
```

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:216]:
```suggestion
to the same class template. It is unspecified, which namespace the underlying class template and its specializations
are defined in.
```

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:222]:
```suggestion
   For a new code, use of the type trait utilities in ``namespace oneapi::dpl`` is highly encouraged. Those
```

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:226]: What's the point of this super generic and kind of vague wording?

I specifically mean "_is one of the above specified or implementation-defined oneDPL execution policy types_".

**Review by rarutyun — COMMENTED** [review]
Overall, looks good to me. I left some small comments that do look as improvements to me. But you might disagree

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:200]: It's a template, `inline` should not be needed. See the standard (https://eel.is/c++draft/execution.syn):
```
namespace std {
  // [[execpol.type]](https://eel.is/c++draft/execpol.type), execution policy type trait
  template<class T> struct is_execution_policy;
  template<class T> constexpr bool [is_execution_policy_v](https://eel.is/c++draft/execution.syn#lib:is_execution_policy_v) = is_execution_policy<T>::value;
}
```

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:208]: Same here.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:226]: Compare with the C++ standard wording:
> `is_execution_policy<T>` is a Cpp17UnaryTypeTrait with a base characteristic of `true_type` if `T` is the type of a standard or implementation-defined execution policy, otherwise `false_type`.

It is assumed that an implementation can define additional policy types, for which the trait also resolves to `true_type`.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:200]: OK, seems `inline` is necessary for C++!7/20. Thanks, I will get it back.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:216]: Spellcheckers do not complain :), so I will keep this intact.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:222]: Made a similar change.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:216]: Sure, this is small suggestion. As far as I understand "_preposition_ + which" is consider to be poorer English. More common use is "bla-bla, _which_ <some words> + <_proposition_ at the end>. Again, not proposing to change that but for my curiosity we could summon a native speaker. @danhoeflinger, could you please share your opinion?

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:226]: I see. Given that this is a specification, we allow the implementations of oneDPL spec to have their own _implementation-defined_ execution policies, if they will.

Thanks.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (rarutyun) [inline source/elements/oneDPL/source/parallel_api/execution_policies.rst:200]:
```suggestion
      inline constexpr bool is_execution_policy_v = oneapi::dpl::is_execution_policy<T>::value;
```

I prefer that order because this is the way it's defined in C++ standard. I don't want to hold this PR because of that issue but if you decide to change it the way I am suggesting, please ping me and I approve this patch once again.

**Review by rarutyun — APPROVED** [review]

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst]: No problem, I will do it. No re-approval is needed, thanks.

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:216]: I searched for clarification on this, and I have found that "preposition + which" is considered not poorer but more formal English.

Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:216]: I agree that "in which" is more formal (but a bit awkward), and "which ... in" is less formal, and a "dangling preposition" https://www.thefreedictionary.com/Dangling-Prepositions.htm which is a classic English grammar issue.

I was trying to rephrase to avoid the problem, but I think it may make it less clear.

"The location of the underlying class template definition and its specializations is left unspecified."

**Review by danhoeflinger — APPROVED** [review]

File: source/elements/oneDPL/source/parallel_api/execution_policies.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst:216]: Thanks Dan, I like the last suggestion, will apply.

Comment (rarutyun) [reply source/elements/oneDPL/source/parallel_api/execution_policies.rst]: > "It is unspecified, which namespace contains the underlying class template definition and its specializations."

I also like that.

> ... not poorer but more formal English.

Thanks for clarification

**Pull request merged**
